### PR TITLE
Excercise 9

### PR DESCRIPTION
**Pull request by stepansojka**

### CODE REVIEW

File: src/modules/crud/crud-saga.js [hunk 1]
```diff
@@ -1,17 +1,30 @@
...
-      };
+      if (isUpdate) {
+        return {
+          effect: updateUser,
+          schema: user
+        };
+      } else {
+        return {
+          effect: saveUser,
+          schema: user
```
Comment (branoholy) [inline src/modules/crud/crud-saga.js]: Not a problem but could be shorter.
```js
return {
  effect: isUpdate ? updateUser : saveUser,
  schema: user
};
```

File: src/modules/entities/entities-selectors.js [hunk 2]
```diff
@@ -1,11 +1,10 @@
 import { createSelector } from "reselect";
 
 const getEntities = state => {
-  //  console.log(state);
   return state.entities;
 };
 
-const getSkills = createSelector(
+export const getNormalizedSkills = createSelector(
```
Comment (branoholy) [inline src/modules/entities/entities-selectors.js]: Skills don't have relationship to other entities so they are not normalized.

File: src/modules/entities/entities-selectors.js [hunk 3]
```diff
@@ -20,9 +19,20 @@ const getNormalizedUserSkills = createSelector(
   entities => entities.userSkills
 );
 
+export const getSkills = createSelector(
```
Comment (branoholy) [inline src/modules/entities/entities-selectors.js]: This selector seems to be useless. Why do you need it?

File: src/modules/users/components/user-create.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+const DumbUserCreate = ({ saveUser }) => <UserCreateForm onSubmit={saveUser} />;
+
+const mapStateToProps = state => ({});
+
+const mapDispatchToProps = {
+  saveUser: UserActionCreators.saveUser
+};
+
+export const UserCreate = connect(
+  mapStateToProps,
```
Comment (branoholy) [inline src/modules/users/components/user-create.js]: You can use `null` or `undefined` if you don't need `mapStateToProps`.

File: src/modules/users/components/user-create.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import React from "react";
+import { connect } from "react-redux";
+
+import { UserCreateForm } from "modules/users/components/user-form";
+import { UserActionCreators } from "modules/users/user-actions";
+
+const DumbUserCreate = ({ saveUser }) => <UserCreateForm onSubmit={saveUser} />;
```
Comment (branoholy) [inline src/modules/users/components/user-create.js:7]: Doesn't matter now but it's more common to use something like `UserCreateComponent`. We actually don't use `Dumb*`. :slightly_smiling_face:

File: src/modules/users/components/user-form.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+const validateFirstName = FormValidations.notEmptyString(
+  "first name cannot be empty"
+);
+const validateLastName = FormValidations.notEmptyString(
+  "last name cannot be empty"
+);
+
+const validateSkills = skills => {
+  if (!Boolean(skills)) {
```
Comment (branoholy) [inline src/modules/users/components/user-form.js]: Not saying this is a problem. Someone likes `Boolean` and I don't have a problem with it. So just FYI, this line is the same as
```js
if (!skills) {
```

I would use the shorter version btw.

File: src/server.js [hunk 7]
```diff
@@ -22,47 +23,82 @@ const reading = {
...
-    (count, user) => (user.firstName === name ? count + 1 : count),
+  Object.keys(users).reduce(
+    (count, userId) => (users[userId].firstName === name ? count + 1 : count),
     1
   );
 
-const createUser = ({ firstName, lastName }) => {
+const createUser = ({ firstName, lastName, skills }) => {
   const regnalNumber = computeRegnalNumber(firstName);
+  const id = `user-${userCount}`;
```
Comment (branoholy) [inline src/server.js]: You have 2 sources of truth for user count. I would rather use `Object.keys(users).length`.

File: src/server.js [hunk 8]
```diff
@@ -22,47 +23,82 @@ const reading = {
...
+
+app.patch("/users/:id", (req, res) => {
+  if (!users[req.params.id]) {
+    res.send(404);
+    return;
+  }
+
+  const update = req.body;
+  const id = req.params.id;
+  const user = users[id];
```
Comment (branoholy) [inline src/server.js]: Both lines for `id` (which can be destructured along with `body`) and `user` can be moved to the beginning and the `if` can be like
```js
if (!user) {
```

File: src/server.js [hunk 9]
```diff
@@ -22,47 +23,82 @@ const reading = {
...
+    (userArray, userId) => [...userArray, users[userId]],
+    []
+  );
+  res.json(u);
+});
+
+app.get("/skills", (req, res) => res.json(allSkills));
+
+app.get("/users/:id", (req, res) => {
+  if (!users[req.params.id]) {
```
Comment (branoholy) [inline src/server.js]: Might be better to save this into a variable because it's used in 2 places.
```js
const user = users[req.params.id];
```

**Review by branoholy — CHANGES_REQUESTED** [review]

File: src/modules/crud/crud-saga.js (hunk 1 above)
Comment (stepansojka) [reply src/modules/crud/crud-saga.js]: makes sense - updated

File: src/modules/entities/entities-selectors.js (hunk 3 above)
Comment (stepansojka) [reply src/modules/entities/entities-selectors.js]: For the user form - the FormFieldSelect component needs an array (methinks) and normalizr stores the skills as object (keys being the skill ids)

File: src/modules/entities/entities-selectors.js (hunk 2 above)
Comment (stepansojka) [reply src/modules/entities/entities-selectors.js]: True. Updated.

File: src/modules/users/components/user-create.js (hunk 4 above)
Comment (stepansojka) [reply src/modules/users/components/user-create.js]: Cheers, I didn't know that - updated.

File: src/modules/users/components/user-form.js (hunk 6 above)
Comment (stepansojka) [reply src/modules/users/components/user-form.js]: Yep the shorter version looks better

File: src/server.js (hunk 7 above)
Comment (stepansojka) [reply src/server.js]: Makes sense - updated

File: src/server.js (hunk 8 above)
Comment (stepansojka) [reply src/server.js]: True - updated the code

File: src/server.js (hunk 9 above)
Comment (stepansojka) [reply src/server.js]: Makes sense - variable created

File: src/modules/users/components/user-create.js (hunk 5 above)
Comment (stepansojka) [reply src/modules/users/components/user-create.js:7]: :) Will rename that in version 2.0 :package:

File: src/modules/entities/entities-selectors.js (hunk 3 above)
Comment (branoholy) [reply src/modules/entities/entities-selectors.js]: The purpose of entity selectors is just denormalization. If you need an array of all skills you should use the `result` of normalization (the array of IDs) and map the IDs to the (denormalized) entities. The `result` is currently stored in the CRUD reducer so you should use it and move this selector to that file.

There is also one technical "problem" with the current implementation. The order of `Object.keys` is not guaranteed and it's common to keep the same order as it was received from the BE (if we don't want to change the order intentionally).

Comment (stepansojka) [reply src/modules/entities/entities-selectors.js]: Yea, I didn't realize that. Updated.

**Review by branoholy — APPROVED** [review]

**Pull request merged**